### PR TITLE
feat: add median implementation 

### PR DESCRIPTION
**Pull request by 0xNeshi**
Original PR #135 by @qalisander 

* add separate quick_sort implementations for each type

* remove duplicated qsort implementations

* add in-place qsort implementation

* add partition function

* rewrite qsort without recursion

* convert quick_sort to macro

* add comprehensive test cases for qsort

* update qsort documentation

* add review fixes

* remove --trace flag from testing pipeline

* Revert "remove --trace flag from testing pipeline"

This reverts commit 9bf706592aba5cae2efb78d9d5c92dd353243c33.

* split macro module

* inline partition macro

* add median implementation

* add more edge case test cases

* fix fmt

* remove quick_sort duplicate

* feat: add median

* docs: update PR

* Route median through u256 workhorse fun (#325)

* fix: fmt

* fix: apply CR comments

* test: add missing test

* fix: macro

* feat: use quickselect

* feat: improve CHANGELOG

* feat: apply review updates

* feat: update comment

---------

Resolves #??? 

#### PR Checklist

- [ ] Tests
- [ ] Documentation
- [ ] Changelog

### CODE REVIEW

**Review by ericnordelo — APPROVED** [review]
LGTM!

**Review by bidzyyys — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,10 @@ The format follows [Keep a Changelog](https://keepachangelog.com/en/1.1.0/)
 
 ### `openzeppelin_math`
 
+#### Added
+
+- Added `vector::median` macro for unsigned integer vectors with rounding for even lengths. Includes `vector::median_u256`. (#135)
```
Comment (0xNeshi) [inline CHANGELOG.md]:
```suggestion
- Added `vector::median` macro for unsigned integer vectors with rounding for even lengths. Includes `vector::median_u256`. (#362)
```

File: math/core/sources/vector.move [hunk 2]
```diff
@@ -194,3 +202,199 @@ public macro fun quick_sort_by<$T>($vec: &mut vector<$T>, $le: |&$T, &$T| -> boo
...
+///
+/// let v = vector[3u64, 1, 4, 1, 5, 9, 2, 6];
+/// let m = vector::median!(&v, rounding::down());
+/// // m == 3
+/// ```
+public macro fun median<$Int>($vec: &vector<$Int>, $rounding_mode: RoundingMode): $Int {
+    let vec = $vec;
+    let vec_u256 = vec.map_ref!(|x| *x as u256);
+    median_u256(vec_u256, $rounding_mode) as $Int
+}
```
[the remark this answers is not in the document]
Comment (0xNeshi) [reply math/core/sources/vector.move:246]: It's a valid `std::vector` macro: https://docs.sui.io/references/framework/sui_std/vector#std_vector_map_ref

**Review by ericnordelo — APPROVED** [review]
